### DISCUSSION TITLE
Order of templates is reversed with nested templates

### DISCUSSION THREAD

**Discussion by lupinitylabs** [alpinejs/alpine#3032] [7. Bugs (archived)]
This issue seems to be related to #2210 / #2209. If two or more `<template>` tags are nested, the output order gets reversed.
#2209 suggests using a `<div>` tag to wrap each nested `<template>` tag, and this works - but I suppose this is not expected behavior?

Here's a jsfiddle demonstrating the issue:
https://jsfiddle.net/8darkhgv/

**Comment by manticorp** [comment]
I too am having this issue.

Wrapping is not an option in my case, as each element needs to be a direct child of the parent.

The jsfiddle shows it perfectly - at the moment, the only solution seems to be to unwrap/flatten the array for the first ```x-for```

**Comment by ekwoka** [comment]
This is just how it's likely got to be, since there is no other way to ensure elements will stay together, as structures like document fragments will be split up once its applied.

**Comment by docdunning** [comment]
The code in the `x-for` directive is way too complicated for me to fathom, but I think the root cause is that it uses `.after` to insert the new elements into the DOM.  This will always put the latest addition at the top of the target element.

So in the jsfiddle example, the first set of numbers is added OK.  Then the second set gets added - again at the top of the target, which is _before_  the first set.  This only happens when the two `x-for` templates are immediately nested - there is only one  target for the new elements.

If the inner `x-for` template is wrapped in another element, then each new set of numbers is inserted at the top of its own wrapper element, and the reversal doesn't happen.

A workaround is to reverse the array/object being iterated, but this is only for the situation where the inner template is directly nested.  See example page below.  I've used the jsfiddle example but with `Alpine.data()` instead of inline `x-data.` 

```
<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <title>Document</title>
  <script defer src="https://unpkg.com/alpinejs@3.x.x/dist/cdn.min.js"></script>
</head>
<body>
<div x-data="numbers">
  <ul>
    Wrong
    <template x-for="sequences in numbers">
      <template x-for="number in sequences">
        <li x-text="number"></li>
      </template>
    </template>
  </ul>
  <ul>
    Workaround
    <template x-for="sequences in reverse(numbers)">
      <template x-for="number in sequences">
        <li x-text="number"></li>
      </template>
    </template>
  </ul>
  <ul>
    Correct
    <template x-for="sequences in numbers">
      <span>
        <template x-for="number in sequences">
          <li x-text="number"></li>
        </template>
      </span>
    </template>
  </ul>
</div>
<script>
document.addEventListener('alpine:init', () => {
  Alpine.data('numbers',() => ({
    numbers: { 1: [1, 2, 3, 4, 5], 2: [6, 7, 8, 9, 10] },
    reverse(arrObj) {
      const keys = Object.keys(arrObj).toReversed()
      return keys.map((item) => arrObj[item])
    },
  }))
})
</script>
</body>
</html>
```

**Reply by ekwoka** [reply]
Basically first outer template is added, and it initializes.

Then second template is added after that element (not after its duplicated elements) which then renders it's children.

I could imagine 2 solutions that aren't too invasive, I think

```
1. Check if template child rendered children and find the last of those
2. Add all the children from an x-for before initializing any of the child trees.
```

2 is probably the smallest code change.

**Reply by lupinitylabs** [reply]
@docdunning I would not suggest to reverse the items iterated, because this will break the second there's a framework fix.

**Reply by docdunning** [reply]
> I would not suggest to reverse the items iterated, because this will break the second there's a framework fix.

I agree!  It's definitely a workaround, and I don't suggest it as a permanent solution.

**Reply by SimoTod** [reply]
> Basically first outer template is added, and it initializes.
> 
> Then second template is added after that element (not after its duplicated elements) which then renders it's children.
> 
> I could imagine 2 solutions that aren't too invasive, I think
> 
> ```
> 1. Check if template child rendered children and find the last of those
> 2. Add all the children from an x-for before initializing any of the child trees.
> ```
> 
> 2 is probably the smallest code change.

You can try to defer https://github.com/alpinejs/alpine/blob/main/packages/alpinejs/src/directives/x-for.js#L210 to a microtask and see how it goes (but I'm not sure if reordering elements will need fixing too).
